### PR TITLE
Created Quote Plugin

### PR DESCRIPTION
**Pull request by landoniwnl**
This plugin will reply with random quotes

### CODE REVIEW

**Comment by Armintron** [comment]
How are users meant to know the set of valid tags?

**Comment by landoniwnl** [comment]
> How are users meant to know the set of valid tags?

They have to check the source code or the api. I want to allow the users to get a list of tags but I'm not sure how to go about implementing it.

**Comment by Armintron** [comment]
> > How are users meant to know the set of valid tags?
> 
> They have to check the source code or the api. I want to allow the users to get a list of tags but I'm not sure how to go about implementing it.

You could just reply with the list of tags when someone tries to use a tag but it isn't valid. "That tag was invalid. Valid tag options are: [valid tag options]"

File: src/app/plugins/quote.plugin.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+      return;
+    }
+
+    if (args.length > 0 && args.length < 2) {
+      if (!this._VALID_TAGS.includes(args[0].toLowerCase())) {
+        await message.reply('Invalid tag, please try again.');
+        return;
+      }
+      url = `${url}?tags=${args[0].toLowerCase()}`;
+    }
```
Comment (tanndlin) [inline src/app/plugins/quote.plugin.ts:74]: Use plugin.validate for validation
[Override this](https://github.com/cs-discord-at-ucf/lion/blob/5ac3e93821667309d981baea9726e837540110a0/src/common/plugin.ts#L35-L41)

File: src/app/plugins/quote.plugin.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+    await this.container.httpService
+      .get(url)
+      .then((response: IHttpResponse) => {
+        const content: string = response.data.content;
+        const author: string = response.data.author;
```
Comment (tanndlin) [inline src/app/plugins/quote.plugin.ts:80]:
```suggestion
        const { content, author } = response.data;
```

File: src/app/plugins/quote.plugin.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        let tag: string =
+          args.length > 0
+            ? args[0].charAt(0).toUpperCase() + args[0].slice(1)
+            : response.data.tags[0].charAt(0).toUpperCase() + response.data.tags[0].slice(1);
```
Comment (tanndlin) [inline src/app/plugins/quote.plugin.ts:84]: Looks like some magic, should prob make its own function

File: src/app/plugins/quote.plugin.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+        if (tag === 'Famous-quotes') tag = 'Famous-Quotes';
```
Comment (tanndlin) [inline src/app/plugins/quote.plugin.ts:86]: Same pattern as line 55, is this necessary?

File: src/app/plugins/quote.plugin.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+  private _createEmbed(content: string, author: string, tag: string) {
+    const embed: MessageEmbed = new MessageEmbed()
+      .setColor('#0099ff')
+      .setTitle(`${tag}`)
+      .setFooter('Pulled from api.quotable.io')
+      .setTimestamp(new Date())
+      .setDescription(
+        `${content}
+
+        - ${author}`
```
Comment (tanndlin) [inline src/app/plugins/quote.plugin.ts:52]: I dont think blank spaces work in embeds

File: src/app/plugins/quote.plugin.ts [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+      );
+
+    if (tag === 'Famous-quotes') embed.setTitle('Famous-Quotes');
+
+    return embed;
+  }
```
Comment (tanndlin) [inline src/app/plugins/quote.plugin.ts:58]:
```suggestion
  private _createEmbed(content: string, author: string, tag: string) {
    return new MessageEmbed()
      .setColor('#0099ff')
      .setTitle(tag === 'Famous-quotes' ? 'Famous-Quotes' : tag)
      .setFooter('Pulled from api.quotable.io')
      .setTimestamp(new Date())
      .setDescription(
        `${content}
        - ${author}`
      );
  }

```

**Review by tanndlin — CHANGES_REQUESTED** [review]
Need to rebase from develop to remove the extra commit :)

File: src/app/plugins/quote.plugin.ts [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+import { MessageEmbed } from 'discord.js';
+
+export default class QuotePlugin extends Plugin {
+  public commandName: string = 'quote';
+  public name: string = 'Quote Plugin';
+  public description: string = 'Prints a random quote';
+  public usage: string = 'quote <tag (optional)>';
+  public override pluginAlias = [];
+  public permission: ChannelType = ChannelType.Public;
+  public override pluginChannelName: string = Constants.Channels.Public.General;
```
Comment (tanndlin) [inline src/app/plugins/quote.plugin.ts:13]: Why only in general

**Comment by tanndlin** [comment]
> > How are users meant to know the set of valid tags?
> 
> They have to check the source code or the api. I want to allow the users to get a list of tags but I'm not sure how to go about implementing it.

You can make a subcommand like
`!quote list`

File: src/app/plugins/quote.plugin.ts (hunk 5 above)
Comment (landoniwnl) [reply src/app/plugins/quote.plugin.ts:52]: It works for me

File: src/app/plugins/quote.plugin.ts [hunk 8]
```diff
@@ -0,0 +1,99 @@
+import { Plugin } from '../../common/plugin';
+import { ChannelType, IContainer, IHttpResponse, IMessage } from '../../common/types';
+import { MessageEmbed } from 'discord.js';
+
+export default class QuotePlugin extends Plugin {
+  public commandName: string = 'quote';
+  public name: string = 'Quote Plugin';
+  public description: string = 'Prints a random quote';
+  public usage: string = 'quote <tag (optional) | list (optional)>';
```
Comment (tanndlin) [inline src/app/plugins/quote.plugin.ts:9]: Usage supports multiple lines for different examples now, separated by newlines
`quote\nquote <tag>\n quote list`

File: src/app/plugins/quote.plugin.ts [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+import { MessageEmbed } from 'discord.js';
+
+export default class QuotePlugin extends Plugin {
+  public commandName: string = 'quote';
+  public name: string = 'Quote Plugin';
+  public description: string = 'Prints a random quote';
+  public usage: string = 'quote <tag (optional) | list (optional)>';
+  public override pluginAlias = [];
+  public permission: ChannelType = ChannelType.Public;
+  public override commandPattern: RegExp = /(list|\w?)/;
```
Comment (tanndlin) [inline src/app/plugins/quote.plugin.ts:12]: don't need this anymore

**Review by tanndlin — COMMENTED** [review]
Few minor things

**Review by tanndlin — APPROVED** [review]

**Pull request merged**